### PR TITLE
Add CatBoost integration

### PR DESCRIPTION
**Pull request by y0z**
## Motivation
- Progress https://github.com/optuna/optuna/issues/4484

## Description of the changes
- Add optuna/integration/catboost.py and its test from optuna/optuna to optuna-integration for isolation of the CatBoost integration

### CODE REVIEW

File: optuna_integration/catboost.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    """Callback for catboost to prune unpromising trials.
+
+    See `the example <https://github.com/optuna/optuna-examples/blob/main/
+    catboost/catboost_pruning.py>`__
+    if you want to add a pruning callback which observes validation accuracy of
+    a CatBoost model.
+
+    .. note::
+        :class:`optuna.TrialPruned` cannot be raised
+        in :meth:`~optuna.integration.CatBoostPruningCallback.after_iteration`
```
Comment (nabenabe0928) [inline optuna_integration/catboost.py:29]:
```suggestion
        in :meth:`~optuna_integration.CatBoostPruningCallback.after_iteration`
```

File: optuna_integration/catboost.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        that is called in CatBoost via ``CatBoostPruningCallback``.
+        You must call :meth:`~optuna.integration.CatBoostPruningCallback.check_pruned`
```
Comment (nabenabe0928) [inline optuna_integration/catboost.py:31]:
```suggestion
        You must call :meth:`~optuna_integration.CatBoostPruningCallback.check_pruned`
```

File: optuna_integration/catboost.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
+from typing import Any
+from typing import Optional
```
Comment (nabenabe0928) [inline optuna_integration/catboost.py:2]:
```suggestion
from __future__ import annotations

from typing import Any
```

File: optuna_integration/catboost.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            for further details.
+        eval_set_index:
+            The index of the target validation dataset.
+            If you set only one ``eval_set``, ``eval_set_index`` is None.
+            If you set multiple datasets as ``eval_set``, the index of ``eval_set`` must be
+            ``eval_set_index``, e.g., ``0`` or ``1`` when ``eval_set`` contains two datasets.
+    """
+
+    def __init__(
+        self, trial: optuna.trial.Trial, metric: str, eval_set_index: Optional[int] = None
```
Comment (nabenabe0928) [inline optuna_integration/catboost.py:58]:
```suggestion
        self, trial: optuna.trial.Trial, metric: str, eval_set_index: int | None = None
```

**Review by nabenabe0928 — COMMENTED** [review]
I left some comments.

**Comment by y0z** [comment]
Thank you for your feedback.
I checked and applied your suggestions.
PTAL.

**Review by nabenabe0928 — APPROVED** [review]
LGTM!

**Pull request merged**